86cutf5yg/renew token

abhijith-biju opened the pull request:
## Overview
This pull request introduces integration of renew-token and logout api.

## Changes Made

### 1. Renew Token funcitonality
- Renew JWT tokens periodically before access token expires.
- Add feature to retry renew-token once more on failure(excluding 401 error).
- Redirect to login modal on 401 responses.

### 2. Logout functionality
- Integrated logout api to revoke tokens for improved safety.

### 3. UI Changes
- Add confirmation alert when closing sign-up modal after making any changes.
- Remove auto-capitalization for button texts. 

## UI Reference
![confirmation alert](https://github.com/nijin-26/bus-ticket-booking-react/assets/146805263/2674f895-b26a-412d-9bd8-194b77309636)

abhijith-biju committed:
feat(auth): add renew token functionality
- add logic inside request interceptor to renew token if access token has expired.

abhijith-biju committed:
feat(renewToken): add renewToken functionality
- add logic in response interceptor to renew token in case of 401 response.
- update renew-token api endpoint to use refreshApi axios instance.
- refactor interceptor addition code.

abhijith-biju committed:
fix(renewToken): fix bug when token expires
- update the failed request retry to use apiClient axios instance.
- module augmented axios library to update AxiosRequestConfig to include custom _retry property.
- update response interceptor to perform only a single retry after renewing tokens.

abhijith-biju committed:
Merge branch develop into 86cuerykj/Auth

abhijith-biju committed:
refactor(renewToken): add toast for token expiry
- update storage.getItem to handle errors during parse.
- update getUserDataFromStorage to use getToken method.

abhijith-biju committed:
feat(auth): integrate logout api
- update imports for FullScreenLoader

abhijith-biju committed:
fix(renewToken): add translation for session timeout toast.
- add navigation to home page after logout.
- move action dispatch from render to effect.

abhijith-biju committed:
fix(requireAuth): remove commented code and add missing dependencies

abhijith-biju committed:
feat(renewToken): add renew-token functionality
- renew-token periodically before accessToken expires
- keep user logged in when loading app if JWT tokens in storage

abhijith-biju committed:
fix(renewToken): clear interval on renewToken failure
- reduce tab heading text size.
- add alert for unsaved data when closing signUp modal.
- changes to logout confirmation modal message.

abhijith-biju committed:
fix(signUpModal): translation bug for unsaved changes alert when closing modal
- remove close authModal on backdrop click

abhijith-biju committed:
refactor(logout): add useLogout custom hook
- clear userData and tokens in case of 401 error.
- notify and keep user logged in case any other error.
- retry renew-token request once more on failure (excluding 401).
- fix translation for cancel booking.

abhijith-biju committed:
refactor(auth): move persistant login logic to a custom hook
- add useLoadInitialAuthState hook

abhijith-biju committed:
Merge branch develop into 86cutf5yg/Renew-Token

abhijith-biju committed:
chore: update changelog

sania-salim commented on src/utils/authUtils/authUtils.ts:
```diff
@@ -1,28 +1,112 @@
+    const remainingTime = getRemainingTimeForExpiry(accessToken);
+    if (remainingTime) {
+        // renew token every 2 minutes before access token expires
+        const intervalForRenew = remainingTime - 120000;
```
What happens when negative intervalForRenew appears?

abhijith-biju replied:
I updated to set a minimum interval of 1 minute.

sania-salim commented on src/hooks/useLogout.ts:
```diff
@@ -0,0 +1,56 @@
+                return;
+            }
+
+            console.error(error);
```
Do we need to log error to console?

abhijith-biju replied:
Removed.

abhijith-biju committed:
fix: resolve MR comments

abhijith-biju committed:
fix(auth): bug when directly loading a protected route
- login modal is displayed when app is loaded directly from a protected route even when logged in.

abhijith-biju committed:
Merge branch develop into 86cutf5yg/Renew-Token

abhijith-biju committed:
fix(signIn): bug where signUp modal shown instead of signIn.
- bug due auth modal retaining selected tab, causing signUp being shown in unauthorized errors.

abhijith-biju committed:
Merge branch develop into 86cutf5yg/Renew-Token

abhijith-biju committed:
fix(auth): store updated tokens to local storage
